backend/src/apiserver/server/util.go

```diff
@@ -216,64 +216,88 @@ func ValidateExperimentResourceReference(resourceManager *resource.ResourceManag
 	return nil
 }
 
-func ValidatePipelineSpec(resourceManager *resource.ResourceManager, spec *api.PipelineSpec) error {
-	if spec == nil || (spec.GetPipelineId() == "" && spec.GetWorkflowManifest() == "") {
-		return util.NewInvalidInputError("Please specify a pipeline by providing a pipeline ID or workflow manifest.")
+func ValidatePipelineSpecAndResourceReferences(resourceManager *resource.ResourceManager, spec *api.PipelineSpec, resourceReferences []*api.ResourceReference) error {
+	// centralize all the nil checks here
+	pipelineId := spec.GetPipelineId()
+	workflowManifest := spec.GetWorkflowManifest()
+	pipelineVersionId := getPipelineVersionIdFromResourceReferences(resourceManager, resourceReferences)
+
+	if err := validateParameters(spec.GetParameters()); err != nil {
+		return err
+	}
+	if workflowManifest != "" {
+		if pipelineId != "" || pipelineVersionId != "" {
+			return util.NewInvalidInputError("Please don't specify a pipeline version or pipeline ID when you specify a workflow manifest.")
+		}
+		if err := validateWorkflowManifest(spec.GetWorkflowManifest()); err != nil {
+			return err
+		}
+	} else {
+		if pipelineId == "" && pipelineVersionId == "" {
+			return util.NewInvalidInputError("Please specify a pipeline by providing a pipeline spec or/and pipeline version.")
+		}
+		if err := validatePipelineId(resourceManager, pipelineId); err != nil {
+			return err
+		}
+		if pipelineVersionId != "" {
+			// verify pipelineVersionId exists
+			pipelineVersion, err := resourceManager.GetPipelineVersion(pipelineVersionId)
```

Comment:
discussion: this isn't a problem with this PR, but I think we should keep it in mind and think about a solution.
Right now, handling a single API request may trigger many resource get requests from mysql:
* once in validation
* once in authorization
* once in actually handling the request

But essentially, they are getting the same information.
Any ideas to remove the redundancy?
(maybe let's create a separate issue for this)